naga/src/back/spv/block.rs

```diff
@@ -293,29 +341,70 @@ impl<'w> BlockContext<'w> {
                         // that actually dereferences the pointer.
                         0
                     }
+                    _ if self.function.spilled_accesses.contains(base) => {
+                        // As far as Naga IR is concerned, this expression does not yield
+                        // a pointer (we just checked, above), but this backend spilled it
+                        // to a temporary variable, so SPIR-V thinks we're accessing it
+                        // via a pointer.
+
+                        // Since the base expression was spilled, mark this access to it
+                        // as spilled, too.
+                        self.function.spilled_accesses.insert(expr_handle);
+                        self.maybe_access_spilled_composite(expr_handle, block, result_type_id)?
+                    }
                     crate::TypeInner::Vector { .. } => {
                         self.write_vector_access(expr_handle, base, index, block)?
                     }
-                    // Only binding arrays in the `Handle` address space will take this
-                    // path, since we handled the `Pointer` case above.
+                    crate::TypeInner::Array { .. } | crate::TypeInner::Matrix { .. } => {
+                        // See if `index` is known at compile time.
+                        match GuardedIndex::from_expression(index, self.ir_function, self.ir_module)
+                        {
+                            GuardedIndex::Known(value) => {
+                                // If `index` is known, we can just use `OpCompositeExtract`.
+                                //
+                                // We never need bounds checks for these cases: everything
+                                // size is statically known and checked in validation.
```

Comment:
For linking purposes so that we remember to update this in the future: This is not spec compliant but we already have https://github.com/gfx-rs/wgpu/issues/4390 tracking this.

Reply:
I went ahead and filed a separate issue specifically for this spot.